javalin/src/main/java/io/javalin/http/staticfiles/PrecompressingResourceHandler.kt

```diff
@@ -0,0 +1,112 @@
+package io.javalin.http.staticfiles
+
+import io.javalin.core.compression.CompressionStrategy
+import io.javalin.core.util.Header
+import io.javalin.http.LeveledBrotliStream
+import io.javalin.http.LeveledGzipStream
+import org.eclipse.jetty.http.MimeTypes
+import org.eclipse.jetty.util.resource.Resource
+import java.io.*
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+
+
+class PrecompressingResourceHandler(val compStrat: CompressionStrategy) {
+
+    val compressedFiles = HashMap<String, ByteArray>()
+
+    companion object {
+        val excludedMimeTypes = setOf(
+                "image/",
+                "audio/",
+                "video/",
+                "application/compress",
+                "application/zip",
+                "application/gzip",
+                "application/bzip2",
+                "application/brotli",
+                "application/x-xz",
+                "application/x-rar-compressed")
+    }
+
+    fun handle(resource: Resource, req: HttpServletRequest, res: HttpServletResponse): Boolean {
+        if (resource.exists() && !resource.isDirectory) {
+            val target = req.getAttribute("jetty-target") as String
+            var acceptCompressType = CompressType.getByAcceptEncoding(req.getHeader(Header.ACCEPT_ENCODING) ?: "")
+            val contentType = MimeTypes.getDefaultMimeByExtension(resource.file.name)//get content type by file extension
+            val resourceCompressible = !excludedMimeType(contentType)
+            if (!resourceCompressible)
+                acceptCompressType = CompressType.NONE
+            val resultByteArray = getStaticResourceByteArray(resource, target, acceptCompressType)
+            res.setContentLength(resultByteArray.size)
+            res.setHeader(Header.CONTENT_TYPE, contentType)
+
+            if (acceptCompressType != CompressType.NONE)
+                res.setHeader(Header.CONTENT_ENCODING, acceptCompressType.typeName)
+
+            val weakETag = resource.weakETag //jetty resource use weakETag too
+            req.getHeader(Header.IF_NONE_MATCH)?.let { etag ->
+                if (etag == weakETag) {
+                    res.status = 304
+                    return true
+                }
+            }
+            res.setHeader(Header.ETAG, weakETag)
+            resultByteArray.inputStream().copyTo(res.outputStream, 2048)
+            res.outputStream.close()
+            return true
+        }
+        return false
+    }
+
+    private fun getStaticResourceByteArray(resource: Resource, target: String, type: CompressType): ByteArray {
+        if (resource.length() > Int.MAX_VALUE)
+            throw RuntimeException("Precompression doesn't support static file size large than 2GB")
+        val key = target + type.extension
+        if (!compressedFiles.containsKey(key)) {
```

Comment:
compute if absent?

Reply:
Yes, it will check the target file exist in HashMap or not.

Reply:
I will add comment in this line.

Reply:
https://www.baeldung.com/java-map-computeifabsent

Reply:
Thank you!
In kotlin it looks like https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/get-or-put.html.
So I modify to this.
```kotlin
return compressedFiles.getOrPut(target + type.extension) { getCompressedByteArray(resource, type) }
```